### PR TITLE
[AMD] Fuse sigmoid + mul attention output gate into single Triton kernel

### PR DESCRIPTION
**Pull request by yichiche**
## Motivation

The Qwen3.5 `AttentionDecoderLayer` applies an output gate after the attention compute as two separate PyTorch elementwise kernels:

```python
gate = torch.sigmoid(gate)        # kernel launch #1
attn_output = attn_output * gate   # kernel launch #2
```

On MI355 (gfx950), HIP kernel launch overhead is ~4 us per kernel. These are tiny pointwise ops (16K elements at batch=4, decode) — purely launch-overhead bound. Two launches cost 8.6 us total where only ~4.5 us of actual work exists.

## Changes

- **New file**: `python/sglang/jit_kernel/triton/sigmoid_gate_mul.py` — a trivial Triton kernel that computes `x * sigmoid(gate)` in a single launch, promoting to float32 for the sigmoid.
- **`qwen3_5.py`**: Replace the two-op gate with the fused kernel, gated behind `_use_aiter` (HIP only). CUDA path unchanged.
- **`qwen3_next.py`**: Same change (identical gate pattern). Also adds `_use_aiter` module-level flag (was missing).

## Performance

**Model**: Qwen3.5-397B-A17B-MXFP4, TP2, MI355x, ISL=8192, OSL=1024, cc=4

### Kernel-level savings

| Metric | Before | After | Savings |
|---|---|---|---|
| Kernels per attention layer | 2 (sigmoid + mul) | 1 (fused Triton) | -1 kernel |
| Time per attention layer | 8.6 us | ~4.5 us | ~4.1 us |
| Total (15 attention layers) | 129 us | ~67 us | **~62 us** |

### E2E benchmark (40 prompts, cc=4, random ISL=8192 OSL=1024)

| Metric | Baseline | After | Delta |
|---|---|---|---|
| Total throughput (tok/s) | 3,152 | 3,398 | +7.8% |
| Output throughput (tok/s) | 314.2 | 308.0 | -2.0% |
| Median TTFT (ms) | 412.5 | 383.5 | -7.0% |
| Median ITL (ms) | 10.86 | 10.84 | -0.2% |
| Median TPOT (ms) | 12.06 | 11.94 | -1.0% |
| Median E2E Latency (ms) | 11,184 | 11,009 | -1.6% |

The kernel-level savings (~62 us) are <1% of iteration time (~10.9 ms), so the ITL improvement is within run-to-run noise. The TTFT and total throughput improvements are also likely noise from this run.

## Accuracy

| Metric | Value |
|---|---|
| GSM8K accuracy (2000 questions) | 0.911 |
| Invalid rate | 0.4% |
| Threshold | 0.88 |
| **Result** | **PASS** |

## Notes

- The same `sigmoid(gate) * x` pattern exists in `afmoe.py` and `bailing_moe_linear.py` — left for a follow-up PR.
- The Triton kernel works on both CUDA and HIP, but is only enabled on HIP via `_use_aiter` to avoid changing the CUDA code path.

---
### CI States

Latest PR Test (Base): :x: [Run #27216465278](https://github.com/sgl-project/sglang/actions/runs/27216465278)
Latest PR Test (Extra): :x: [Run #27216464940](https://github.com/sgl-project/sglang/actions/runs/27216464940)

### CODE REVIEW

File: python/sglang/jit_kernel/tests/test_sigmoid_gate_mul.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
+import pytest
+import torch
+
+from sglang.test.ci.ci_register import register_cuda_ci
+
+register_cuda_ci(est_time=4, suite="base-b-kernel-unit-1-gpu-large")
```
Comment (bingxche) [inline python/sglang/jit_kernel/tests/test_sigmoid_gate_mul.py:8]: Please also register AMD CI, thanks!

Comment (yichiche) [reply python/sglang/jit_kernel/tests/test_sigmoid_gate_mul.py:8]: Done, thanks for the reminder

File: python/sglang/srt/models/qwen3_5.py [hunk 2]
```diff
@@ -930,8 +930,15 @@ def self_attention(
         attn_output = self.attn(q, k, v, forward_batch)
 
         if self.attn_output_gate:
-            gate = torch.sigmoid(gate)
-            attn_output = attn_output * gate
+            if _use_aiter:
```
Comment (sogalin) [inline python/sglang/srt/models/qwen3_5.py]: Is this related to aiter, if not can we use _is_hip?

File: python/sglang/srt/models/qwen3_next.py [hunk 3]
```diff
@@ -819,8 +823,15 @@ def self_attention(
         attn_output = self.attn(q, k, v, forward_batch)
 
         if self.attn_output_gate:
-            gate = torch.sigmoid(gate)
-            attn_output = attn_output * gate
+            if _use_aiter:
```
Comment (sogalin) [inline python/sglang/srt/models/qwen3_next.py]: Is this related to aiter, if not can we use _is_hip?

Comment (yichiche) [reply python/sglang/srt/models/qwen3_next.py]: Yeah, we use _is_hip instead since this is not dependent on aiter library.

**Comment by yichiche** [comment]
@amd-bot ci-status

**Comment by sogalin** [comment]
@yichiche , The changes look good to me. Could you also provide the results for the full GSM8K test set (1,319 samples)?

**Comment by yichiche** [comment]
> @yichiche , The changes look good to me. Could you also provide the results for the full GSM8K test set (1,319 samples)?

root@smci355-ccs-aus-m15-21:/sgl-workspace/sglang# python3 benchmark/gsm8k/bench_sglang.py --num-questions 2000 --parallel 1000 --num-shots 5 --port 8000
Downloading from https://raw.githubusercontent.com/openai/grade-school-math/master/grade_school_math/data/test.jsonl to /tmp/test.jsonl
/tmp/test.jsonl: 732kB [00:00, 8.21MB/s]                                                                                             
100%|████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [01:42<00:00, 12.85it/s]
Accuracy: 0.911
Invalid: 0.004
Latency: 102.640 s
Output throughput: 2216.560 token/s

**Review by sogalin — APPROVED** [review]

**Comment by sogalin** [comment]
> > @yichiche , The changes look good to me. Could you also provide the results for the full GSM8K test set (1,319 samples)?
> 
> root@smci355-ccs-aus-m15-21:/sgl-workspace/sglang# python3 benchmark/gsm8k/bench_sglang.py --num-questions 2000 --parallel 1000 --num-shots 5 --port 8000 Downloading from https://raw.githubusercontent.com/openai/grade-school-math/master/grade_school_math/data/test.jsonl to /tmp/test.jsonl /tmp/test.jsonl: 732kB [00:00, 8.21MB/s] 100%|████████████████████████████████████████████████████████████████████████████████████████████| 1319/1319 [01:42<00:00, 12.85it/s] Accuracy: 0.911 Invalid: 0.004 Latency: 102.640 s Output throughput: 2216.560 token/s

Thanks, it looks good to me now.

**Review by HaiShaw — APPROVED** [review]
